File: data-structure-and-algorithm/cpp/validate-binary-search-tree.cpp
// 98. Validate Binary Search Tree
// Difficulty: Medium

// Given a binary tree, determine if it is a valid binary search tree (BST).

// Assume a BST is defined as follows:

// The left subtree of a node contains only nodes with keys less than the node's key.
// The right subtree of a node contains only nodes with keys greater than the node's key.
// Both the left and right subtrees must also be binary search trees.
// Example 1:
//     2
//    / \
//   1   3
// Binary tree [2,1,3], return true.
// Example 2:
//     1
//    / \
//   2   3
// Binary tree [1,2,3], return false.

// Time:  O(n)
// Space: O(h)

class Solution {
public:
	bool isValidBST(TreeNode* root) {
		return isValidBST(root, INT_MIN, INT_MAX);
	}

private:
	bool isValidBST(TreeNode* root, int lower, int upper) {
		if (!root) {
			return true;
		}
		if (root->val < lower || root->val > upper) {
			return false;
		}
		return isValidBST(root->left, lower, root->val) && isValidBST(root->right, root->val, upper);
  }
};

File: data-structure-and-algorithm/java/validate-binary-search-tree.java
/**
 * Validate Binary Search Tree
 * 
 * Given a binary tree, determine if it is a valid binary search tree (BST).
 * 
 * Assume a BST is defined as follows:
 * 
 * The left subtree of a node contains only nodes with keys less than the node's key.
 * The right subtree of a node contains only nodes with keys greater than the node's key.
 * Both the left and right subtrees must also be binary search trees.
 * Example 1:
 * 
 * Input:
 *     2
 *    / \
 *   1   3
 * Output: true
 * Example 2:
 * 
 *     5
 *    / \
 *   1   4
 *      / \
 *     3   6
 * Output: false
 * Explanation: The input is: [5,1,4,null,null,3,6]. The root node's value
 *              is 5 but its right child's value is 4.
 */

public class Solution {
    public boolean isValidBST(TreeNode root) {
        return isValidBSTHelper(root, Integer.MIN_VALUE, Integer.MAX_VALUE);
    }

    private boolean isValidBSTHelper(TreeNode node, int min, int max) {
        if (node == null) {
            return true;
        }

        if (node.val > max || node.val < min) {
            return false;
        }

        return isValidBSTHelper(node.left, min, node.val) && isValidBSTHelper(node.right, node.val, max);
    }
}

public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) {
        val = x;
    }
}
